rules/prefer-error-cause.js

```diff
@@ -0,0 +1,388 @@
+		// try {} catch { const err = new Error; try {} catch { throw err; } }
+		if (_.isEmpty(thrownErrorDeclarators)) {
+			reportCannotBeFixed(node, context);
+			return;
+		}
+
+		// try {} catch (err) { let e1 = new Error('oops'); e1 = new Error('oops'); throw error;}
+
+		for (const thrownErrorDeclarator of thrownErrorDeclarators) {
+			statementToFix = thrownErrorDeclarator;
+
+			let targetArguments;
+			if (thrownErrorDeclarator.type === 'VariableDeclarator') {
+				targetArguments = thrownErrorDeclarator.init.arguments;
+			} else if (thrownErrorDeclarator.type === 'AssignmentExpression') {
+				targetArguments = thrownErrorDeclarator.right.arguments;
+			}
+
+			errorConstructorLastArgument = peekTop(targetArguments);
+
+			if (!errorConstructorLastArgument) {
+				reportCannotBeFixed(node, context);
+				return;
+			}
+
+			result.push({
+				node: errorConstructorLastArgument,
+				statementToFix,
+			});
+		}
+	}
+
+	return result;
+};
+
+const handleCatchBlock = ({node, context, statements, parameter}) => {
+	const throwStatement = statements.find(statement => statement.type === 'ThrowStatement');
+
+	// Filter blocks not having throw statement.
+	// try {} catch (error) {}
+	// promise.catch(error => {})
+	if (!throwStatement) {
+		return;
+	}
+
+	const errorArgumentIdentifier = parameter?.name;
+	const throwStatementArgument = throwStatement.argument;
+
+	// Filter block having valid cause property
+	// try {} catch (err) { throw new Error('oops', {cause: err})}
+	// promise.catch(err => { throw new Error('oops', {cause: err})
+	if (errorArgumentIdentifier && errorArgumentIdentifier === throwStatementArgument.name) {
+		return;
+	}
+
+	// Report none identifier parameter
+	// try {} catch ({error}) {}
+	// promise.catch({error} => {})
+	if (parameter && parameter.type !== 'Identifier') {
+		context.report({
+			node,
+			messageId: ERROR,
+		});
+		return;
+	}
+
+	const hasValidCauseProperty = generateCauseInspector(property => property.value.name === errorArgumentIdentifier);
+	const hasInvalidCauseProperty = generateCauseInspector(property => property.value.name !== errorArgumentIdentifier);
+
+	const nodesToFixCandidates = getAllNodesToFix({node, context, throwStatement});
+	if (_.isEmpty(nodesToFixCandidates)) {
+		return;
+	}
+
+	const nodesToFix = nodesToFixCandidates.filter(({node}) =>
+		!hasValidCauseProperty(node),
+	);
+
+	if (!_.isEmpty(nodesToFix)) {
+		for (const nodeToFix of nodesToFix) {
+			// Prevent repetitive fixing
+			if (hasValidCauseProperty(nodeToFix.node)) {
+				continue;
+			}
+
+			context.report({
+				node,
+				messageId: ERROR,
+				* fix(fixer) {
+					const result = fix({
+						fixer,
+						node,
+						context,
+						statementToFix: nodeToFix.statementToFix,
+						errorConstructorLastArgument: nodeToFix.node,
+						errorArgumentIdentifier,
+						isCauseNameValid: hasInvalidCauseProperty(nodeToFix.node),
+					});
+
+					if (result) {
+						yield result;
+					}
```

Comment:
return the result directly, the fix function don't has to be a generator function.